jaxmarl/environments/hanabi/hanabi.py

```diff
@@ -241,63 +241,57 @@ def get_legal_moves(self, state: State) -> chex.Array:
 
         @partial(jax.vmap, in_axes=[0, None])
         def _legal_moves(aidx: int, state: State) -> chex.Array:
-            """
-            Legal moves encoding in order:
-            - discard for all cards in hand
-            - play for all cards in hand
-            - hint for all colors and ranks for all other players
-            """
-            move_idx = 0
+            # all moves are illegal in the beginning
             legal_moves = jnp.zeros(self.num_moves)
-            hands = state.player_hands
-            info_tokens = state.info_tokens
-            # discard legal when discard tokens are not full
-            is_not_max_info_tokens = jnp.sum(state.info_tokens) < self.max_info_tokens
-            legal_moves = legal_moves.at[move_idx : move_idx + self.hand_size].set(
-                is_not_max_info_tokens
-            )
-            move_idx += self.hand_size
-            # play moves always legal
-            legal_moves = legal_moves.at[move_idx : move_idx + self.hand_size].set(1)
-            move_idx += self.hand_size
-            # hints depend on other player cards
-            other_hands = jnp.delete(hands, aidx, axis=0, assume_unique_indices=True)
-
-            def _get_hints_for_hand(carry, unused):
-                """Generates valid hints given hand"""
-                aidx, other_hands = carry
-                hand = other_hands[aidx]
-
-                # get occurrences of each card
-                card_counts = jnp.sum(hand, axis=0)
-                # get occurrences of each color
-                color_counts = jnp.sum(card_counts, axis=1)
-                # get occurrences of each rank
-                rank_counts = jnp.sum(card_counts, axis=0)
-                # check which colors/ranks in hand
-                colors_present = jnp.where(color_counts > 0, 1, 0)
-                ranks_present = jnp.where(rank_counts > 0, 1, 0)
-
-                valid_hints = jnp.concatenate([colors_present, ranks_present])
-                carry = (aidx + 1, other_hands)
-
-                return carry, valid_hints
-
-            _, valid_hints = lax.scan(
-                _get_hints_for_hand, (0, other_hands), None, self.num_agents - 1
+            all_player_hands = state.player_hands  # (num_players, hand_size, num_colors, num_ranks)
+
+            # first get all cards one is holding since only these are legally playable
+            my_hand = all_player_hands.at[aidx].get()  # (hand_size, num_colors, num_ranks)
+            holding_cards_idx = jax.vmap(lambda c: jnp.any(c))(my_hand)
```

Comment:
I don't think we need vmap here, but it seemed easier to read this way. Lmk if it needs changing.